catalog/IndexScheme.hpp

```diff
@@ -0,0 +1,185 @@
+
+/** \addtogroup Catalog
+ *  @{
+ */
+
+/**
+ * @brief The Index Scheme class which stores the information about
+ *        various indicies defined for a particular relation.
+ **/
+class IndexScheme {
+ public:
+  /**
+   * @brief Constructor.
+   **/
+  IndexScheme() {
+  }
+
+  /**
+   * @brief Virtual destructor.
+   **/
+  ~IndexScheme() {
+  }
+
+  /**
+   * @brief Reconstruct an Index Scheme from its serialized
+   *        Protocol Buffer form.
+   *
+   * @param proto The Protocol Buffer serialization of a Index Scheme,
+   *              previously produced by getProto().
+   * @return The deserialied index scheme object.
+   **/
+  static IndexScheme* ReconstructFromProto(const serialization::IndexScheme &proto);
+
+  /**
+   * @brief Check whether a serialization::IndexScheme is fully-formed and
+   *        all parts are valid.
+   *
+   * @param proto A serialized Protocol Buffer representation of a
+   *              IndexScheme, originally generated by getProto().
+   * @return Whether proto is fully-formed and valid.
+   **/
+  static bool ProtoIsValid(const serialization::IndexScheme &proto);
+
+  /**
+   * @brief Serialize the Index Scheme as Protocol Buffer.
+   *
+   * @return The Protocol Buffer representation of Index Scheme.
+   **/
+  serialization::IndexScheme getProto() const;
+
+  /**
+   * @brief Get the number of indices for the relation.
+   *
+   * @return The number of indices defined for the relation.
+   **/
+  inline const std::size_t getNumIndices() const {
+    return index_map_.size();
+  }
+
+  /**
+   * @brief Check whether an index with the given exists or not.
+   *
+   * @param index_name Name of the index to be checked.
+   * @return Whether the index exists or not.
+   **/
+  bool hasIndexWithName(const std::string &index_name) const {
+    return index_map_.find(index_name) != index_map_.end();
+  }
+
+  /**
+   * @brief Check whether an index with the given description
+   *        containing the same attribute id and index type
+   *        exists or not in the index map.
+   *
+   * @param index_descripton Index Description to check against.
+   * @return Whether a similar index description was already defined or not.
+   **/
+  bool hasIndexWithDescription(const IndexSubBlockDescription &index_description) const {
+    // Iterate through every vector of descriptions corresponding to all keys in the index map.
+    for (auto cit = index_map_.cbegin(); cit != index_map_.cend(); ++cit) {
+      std::vector<IndexSubBlockDescription> index_descriptions = cit->second;
+      for (auto index_description_itr = index_descriptions.cbegin();
+           index_description_itr != index_descriptions.cend();
+           ++index_description_itr) {
+        const IndexSubBlockDescription &stored_description = *index_description_itr;
+        // Check if the stored description matches as the given description.
+        if (areIndexDescriptionsSimilar(stored_description, index_description)) {
```

Comment:
Use `*index_description_itr` directly, and remove above definition for `stored_description`.

Rename `index_description_itr` to `description_it`?